File: core/recc/struct/permission.py
# -*- coding: utf-8 -*-

from typing import Optional, Any
from datetime import datetime
from recc.struct._structure import _Structure


class Permission(_Structure):
    def __init__(
        self,
        uid: Optional[int] = None,
        name: Optional[str] = None,
        description: Optional[str] = None,
        extra: Optional[Any] = None,
        r_layout: Optional[bool] = None,
        w_layout: Optional[bool] = None,
        r_storage: Optional[bool] = None,
        w_storage: Optional[bool] = None,
        r_manager: Optional[bool] = None,
        w_manager: Optional[bool] = None,
        r_graph: Optional[bool] = None,
        w_graph: Optional[bool] = None,
        r_member: Optional[bool] = None,
        w_member: Optional[bool] = None,
        r_setting: Optional[bool] = None,
        w_setting: Optional[bool] = None,
        created_at: Optional[datetime] = None,
        updated_at: Optional[datetime] = None,
        **kwargs,
    ):
        self.uid = uid
        self.name = name
        self.description = description
        self.extra = extra
        self.r_layout = r_layout
        self.w_layout = w_layout
        self.r_storage = r_storage
        self.w_storage = w_storage
        self.r_manager = r_manager
        self.w_manager = w_manager
        self.r_graph = r_graph
        self.w_graph = w_graph
        self.r_member = r_member
        self.w_member = w_member
        self.r_setting = r_setting
        self.w_setting = w_setting
        self.created_at = created_at
        self.updated_at = updated_at
        self.kwargs = kwargs

File: fe/src/packet/permission.ts
export interface RawPermission {
    r_layout: boolean;
    w_layout: boolean;
    r_storage: boolean;
    w_storage: boolean;
    r_manager: boolean;
    w_manager: boolean;
    r_graph: boolean;
    w_graph: boolean;
    r_member: boolean;
    w_member: boolean;
    r_setting: boolean;
    w_setting: boolean;
    is_admin: boolean;
    features: Array<string>;
    extra: any;
}

export interface PermissionA {
    slug: string;
    name?: string;
    description?: string;
    features?: Array<string>;
    extra?: any;
    r_layout?: boolean;
    w_layout?: boolean;
    r_storage?: boolean;
    w_storage?: boolean;
    r_manager?: boolean;
    w_manager?: boolean;
    r_graph?: boolean;
    w_graph?: boolean;
    r_member?: boolean;
    w_member?: boolean;
    r_setting?: boolean;
    w_setting?: boolean;
    hidden?: boolean;
    lock?: boolean;
    created_at?: string;
    updated_at?: string;
}

export interface CreatePermissionQ {
    slug: string;
    name?: string;
    description?: string;
    features?: Array<string>;
    extra?: any;
    r_layout?: boolean;
    w_layout?: boolean;
    r_storage?: boolean;
    w_storage?: boolean;
    r_manager?: boolean;
    w_manager?: boolean;
    r_graph?: boolean;
    w_graph?: boolean;
    r_member?: boolean;
    w_member?: boolean;
    r_setting?: boolean;
    w_setting?: boolean;
    hidden?: boolean;
    lock?: boolean;
}

export interface UpdatePermissionQ {
    slug?: string;
    name?: string;
    description?: string;
    features?: Array<string>;
    extra?: any;
    r_layout?: boolean;
    w_layout?: boolean;
    r_storage?: boolean;
    w_storage?: boolean;
    r_manager?: boolean;
    w_manager?: boolean;
    r_graph?: boolean;
    w_graph?: boolean;
    r_member?: boolean;
    w_member?: boolean;
    r_setting?: boolean;
    w_setting?: boolean;
    hidden?: boolean;
    lock?: boolean;
}

export const PERMISSION_SLUG_GUEST = 'guest';
export const PERMISSION_SLUG_REPORTER = 'reporter';
export const PERMISSION_SLUG_OPERATOR = 'operator';
export const PERMISSION_SLUG_MAINTAINER = 'maintainer';
export const PERMISSION_SLUG_OWNER = 'owner';
